### ISSUE TITLE
No matching documents when trying to filter on boolean

### ISSUE THREAD

**Issue by griseau** [vespa-engine/vespa#11884]
Hi, 

When trying to filter documents on a boolean field equals to true I don't have any matching docs, but it works if this boolean is equals to false. 

I'm using Vespa: 7.145.41

In my searchdefinitions the field is 
`field expert type bool {
indexing: summary | attribute
} `

When running this query : `select * from sources post where expert=true;`

I have the following trace `{
                                "message": "sc0.num0 search to dispatch: query=[expert:true] timeout=493ms offset=0 hits=10 groupingSessionCache=true sessionId=container.container.0.1579684136189.1515.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": "Current state of query tree: BOOL[connectedItem=null connectivity=0.0 creator=ORIG explicitSignificance=false index=\"expert\" isRanked=true origin=\"(0 4)\" significance=0.0 uniqueID=1 usePositionData=true weight=100]{\n  \"true\"\n}\n"
                            }`

Thanks !

**Comment by jobergum** [comment]
Thank you for your detailed issue @griseau. Is expert an imported field from a parent document type?

**Comment by griseau** [comment]
No, I have only one document type, and expert belongs to it

**Comment by baldersheim** [comment]
Could you add the full trace when running both queries with "&tracelevel=1&explainlevel=1" ?
If you have more than one backend node you should also add "&searchPath=1/".

**Comment by griseau** [comment]
I only have one backend node at the moment. Trace of `select * from sources post where expert=true;` is : 
`trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[expert:true] timeout=499ms offset=0 hits=10 groupingSessionCache=true sessionId=container.container.0.1579697740322.1519.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time_relative": "1970-01-03 03:46:59.614 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.254199,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::EmptyBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": true,
                                                        "estHits": 0,
                                                        "cost_tier": 1,
                                                        "tree_size": 1,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 0
                                                }
                                            },
                                            {
                                                "timestamp_ms": 11.984468,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.285334,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.889735,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.94672,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.959266,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.963061,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.969805,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 11.971411,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 12.032708
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 0 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    }`

**Comment by baldersheim** [comment]
I also need the same trace for the query where 'expert=false'.

**Comment by griseau** [comment]
Trace of `elect * from sources post where expert=true;` is : 
`"trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[expert:false] timeout=498ms offset=0 hits=10 groupingSessionCache=true sessionId=container.container.0.1579698370855.1520.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time_relative": "1970-01-03 03:57:30.147 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.392736,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::AndBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": false,
                                                        "estHits": 13851732,
                                                        "cost_tier": 1,
                                                        "tree_size": 3,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 13851732,
                                                    "children": {
                                                        "[type]": "std::vector",
                                                        "[0]": {
                                                            "[type]": "search::(anonymous namespace)::AttributeFieldBlueprint",
                                                            "isTermLike": true,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732,
                                                            "fields": {
                                                                "[type]": "FieldList",
                                                                "[0]": {
                                                                    "[type]": "Field",
                                                                    "fieldId": 10,
                                                                    "handle": 0,
                                                                    "isFilter": false
                                                                }
                                                            },
                                                            "attribute": "expert"
                                                        },
                                                        "[1]": {
                                                            "[type]": "proton::documentmetastore::(anonymous namespace)::WhiteListBlueprint",
                                                            "isTermLike": false,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732
                                                        }
                                                    }
                                                }
                                            },
                                            {
                                                "timestamp_ms": 267.486342,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.473135,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.662917,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 265.992326,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 266.277462,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 266.979536,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 267.452243,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 267.457908,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 267.629715
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (10 of total 8496667 hits)"
                            },
                            {
                                "message": "sc0.num0 fill to dispatch: query=[expert:false] timeout=498ms offset=0 hits=10 groupingSessionCache=true sessionId=container.container.0.1579698370855.1520.default grouping=0 :  restrict=[post] summary=[null]"
                            }
                        ]
                    }
                ]
            }
        ]
    }`

**Comment by baldersheim** [comment]
Could you attach the rest of the response you get when running with 'expert=false'. I do not need the documents themselves, but I need some of the meta information.

**Comment by griseau** [comment]
`{
    "trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[expert:false] timeout=500ms offset=0 hits=0 groupingSessionCache=true sessionId=container.container.0.1579703155827.1525.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time_relative": "1970-01-03 05:17:15.119 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.205341,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::AndBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": false,
                                                        "estHits": 13851732,
                                                        "cost_tier": 1,
                                                        "tree_size": 3,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 13851732,
                                                    "children": {
                                                        "[type]": "std::vector",
                                                        "[0]": {
                                                            "[type]": "search::(anonymous namespace)::AttributeFieldBlueprint",
                                                            "isTermLike": true,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732,
                                                            "fields": {
                                                                "[type]": "FieldList",
                                                                "[0]": {
                                                                    "[type]": "Field",
                                                                    "fieldId": 10,
                                                                    "handle": 0,
                                                                    "isFilter": false
                                                                }
                                                            },
                                                            "attribute": "expert"
                                                        },
                                                        "[1]": {
                                                            "[type]": "proton::documentmetastore::(anonymous namespace)::WhiteListBlueprint",
                                                            "isTermLike": false,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732
                                                        }
                                                    }
                                                }
                                            },
                                            {
                                                "timestamp_ms": 106.218138,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.278972,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.316438,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 105.260386,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 105.27784,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 105.975847,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 106.187558,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 106.191385,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 106.27172
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 12570543 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    },
    "root": {
        "id": "toplevel",
        "relevance": 1,
        "fields": {
            "totalCount": 12570543
        },
        "coverage": {
            "coverage": 100,
            "documents": 13851731,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        }
    }
}`

Let me know if you need anything else

**Comment by baldersheim** [comment]
Could you also run the query with 'sddocname contains "post"', so I can see how a it behaves with a different query operator. And add the same output as you just did.

**Comment by griseau** [comment]
Trace for select * from sources post where sddocname contains 'post'; : 
`{
    "trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[[documentmetastore]:*post*] timeout=500ms offset=0 hits=0 groupingSessionCache=true sessionId=container.container.0.1579704209035.1531.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time_relative": "1970-01-03 05:34:48.549 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.24843,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::AndBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": false,
                                                        "estHits": 13851732,
                                                        "cost_tier": 1,
                                                        "tree_size": 3,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 13851732,
                                                    "children": {
                                                        "[type]": "std::vector",
                                                        "[0]": {
                                                            "[type]": "search::(anonymous namespace)::AttributeFieldBlueprint",
                                                            "isTermLike": true,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732,
                                                            "fields": {
                                                                "[type]": "FieldList",
                                                                "[0]": {
                                                                    "[type]": "Field",
                                                                    "fieldId": 16,
                                                                    "handle": 0,
                                                                    "isFilter": true
                                                                }
                                                            },
                                                            "attribute": "[documentmetastore]"
                                                        },
                                                        "[1]": {
                                                            "[type]": "proton::documentmetastore::(anonymous namespace)::WhiteListBlueprint",
                                                            "isTermLike": false,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 13851732,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 13851732
                                                        }
                                                    }
                                                }
                                            },
                                            {
                                                "timestamp_ms": 240.089555,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.284163,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.315124,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 239.155307,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 239.171765,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 239.884659,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 240.062614,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 240.065965,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 240.147483
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 13851731 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    },
    "root": {
        "id": "toplevel",
        "relevance": 1,
        "fields": {
            "totalCount": 13851731
        },
        "coverage": {
            "coverage": 100,
            "documents": 13851731,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        }
    }
}`

Also, I tried on a new machine with the same search definition. If I have only documents where the field expert=true, then the query works on this machine! I'll try to add documents with the boolean expert=false to this engine and see if I have the same output

**Comment by baldersheim** [comment]
All our system tests pass. I have not been able to find any issue here. Yet, there must be one.
In the case where you search for expert=true, it makes an incorrect conclusion that there are actually none. According to the results for the sddocname query which return all documents, and the results for expert=false, there should be 13851731 - 12570543 = 1.3M docs.

So I have to ask some questions here to narrow down what could be the causing this.
1 - Did you feed your documents ? Did you use updates, or put operations ?
2 - Have you set visibility-delay in services.xml to a non-zero value ?
3 - Is there concurrent feed to the system, while you do the queries ?
4 - Have you restarted the application after you fed the documents ?
      If not, could you restart it and redo the query.

There was recently done a change in how we estimate number of hits upfront in the query planning phase affecting boolean fields. A computation was moved from query time to feed time. It would be good if you could verify if this is still an issue in 7.164 that contains this change.

**Comment by griseau** [comment]
1 - I fed the documents using the jar with put operations
2 - I've tried both. Once visibility delay was set to 300, and another time it wasn't set. Output was the same
3 - Maybe we were doing some queries during feeding but when I tried to query with boolean we were not feeding any document
4 - Can't remember if I did restart the application, so I'll try

Not sure to have enough time to do the tests tomorrow but I'll do them asap and get back to you. 

Thanks for your support !

**Comment by baldersheim** [comment]
According to the first trace with expert=true it spends around 10ms for query planning. That sound like a very long time. What hardware are you running on ? Some of the code uses special instructions for bit counting. Could you do 'cat /proc/cpuinfo' ?

**Comment by baldersheim** [comment]
Since I was not able to reproduce here on trunk, I had another look at the differences between 7.145 and trunk and noticed a change that could explain what you are seeing. In #11879 there was introduced an invalidation of the cached bit count for the bool attributes on startup. That invalidation is missing on 7.145.
This will cause it to incorrectly estimate 0 hits in boolean fields until a document is fed. If you feed one document or one update to the expert field this incorrectly cached value will be invalidated and recomputed correctly upon next query.
This issue was fixed in 7.165. So I advise an upgrade to latest released version.

**Comment by griseau** [comment]
I upgraded to 7.165 and it works ! 

Thanks for your support

**Comment by griseau** [comment]
Hi @baldersheim

Looks like it's not working anymore (using Vespa 7.165).  Could you please reopen this issue ? 

I'm using a 8 cores machine and below is the output of one cpu info : 

processor	: 7
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU @ 2.30GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2300.000
cache size	: 46080 KB
physical id	: 0
siblings	: 8
core id		: 3
cpu cores	: 4
apicid		: 7
initial apicid	: 7
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit
bogomips	: 4600.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:

**Comment by baldersheim** [comment]
hmm, what happened between working and then not working ? Was there any feed or any restart ?
Could you provide the 3 different traces you provided earlier on with "&explainlevel=1&tracelevel=1" ?
 With 'expert = true', 'expert = false' and "sddocname contains 'post'".

**Comment by griseau** [comment]
I restarted Vespa and fed more documents with expert=false. 

Trace for expert = true : 
```
{
    "trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[expert:true] timeout=19999ms offset=0 hits=0 groupingSessionCache=true sessionId=container.container.0.1580203337821.4385.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time": "2020-01-28 09:22:17.822 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.209704,
                                                "event": "MTF: Start"
                                            },
                                            {
                                                "timestamp_ms": 0.215763,
                                                "event": "MTF: Build query"
                                            },
                                            {
                                                "timestamp_ms": 0.236607,
                                                "event": "MTF: reserve handles"
                                            },
                                            {
                                                "timestamp_ms": 0.373717,
                                                "event": "MTF: Fetch Postings"
                                            },
                                            {
                                                "timestamp_ms": 0.377084,
                                                "event": "MTF: prepareSharedState"
                                            },
                                            {
                                                "timestamp_ms": 0.403251,
                                                "event": "MTF: Complete"
                                            },
                                            {
                                                "timestamp_ms": 0.404808,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::EmptyBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": true,
                                                        "estHits": 0,
                                                        "cost_tier": 1,
                                                        "tree_size": 1,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 0
                                                }
                                            },
                                            {
                                                "timestamp_ms": 0.62757,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.495928,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.59449,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.603252,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.609906,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.613078,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.619073,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.620016,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 0.665043
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 0 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    },
    "root": {
        "id": "toplevel",
        "relevance": 1,
        "fields": {
            "totalCount": 0
        },
        "coverage": {
            "coverage": 100,
            "documents": 32844580,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        }
    }
}
```

Trace for expert = false
```
{
    "trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[expert:false] timeout=19999ms offset=0 hits=0 groupingSessionCache=true sessionId=container.container.0.1580203392416.4386.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time": "2020-01-28 09:23:12.416 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.209501,
                                                "event": "MTF: Start"
                                            },
                                            {
                                                "timestamp_ms": 0.214526,
                                                "event": "MTF: Build query"
                                            },
                                            {
                                                "timestamp_ms": 0.232796,
                                                "event": "MTF: reserve handles"
                                            },
                                            {
                                                "timestamp_ms": 0.284344,
                                                "event": "MTF: Fetch Postings"
                                            },
                                            {
                                                "timestamp_ms": 0.287319,
                                                "event": "MTF: prepareSharedState"
                                            },
                                            {
                                                "timestamp_ms": 0.298742,
                                                "event": "MTF: Complete"
                                            },
                                            {
                                                "timestamp_ms": 0.30027,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::AndBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": false,
                                                        "estHits": 32844581,
                                                        "cost_tier": 1,
                                                        "tree_size": 3,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 32844581,
                                                    "children": {
                                                        "[type]": "std::vector",
                                                        "[0]": {
                                                            "[type]": "search::(anonymous namespace)::AttributeFieldBlueprint",
                                                            "isTermLike": true,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 32844581,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 32844581,
                                                            "fields": {
                                                                "[type]": "FieldList",
                                                                "[0]": {
                                                                    "[type]": "Field",
                                                                    "fieldId": 10,
                                                                    "handle": 0,
                                                                    "isFilter": false
                                                                }
                                                            },
                                                            "attribute": "expert"
                                                        },
                                                        "[1]": {
                                                            "[type]": "proton::documentmetastore::(anonymous namespace)::WhiteListBlueprint",
                                                            "isTermLike": false,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 32844581,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 32844581
                                                        }
                                                    }
                                                }
                                            },
                                            {
                                                "timestamp_ms": 290.007896,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.580712,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.649611,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 287.587086,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 287.606774,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 289.59978,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 289.978076,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 289.980866,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 290.088815
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 31564042 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    },
    "root": {
        "id": "toplevel",
        "relevance": 1,
        "fields": {
            "totalCount": 31564042
        },
        "coverage": {
            "coverage": 100,
            "documents": 32844580,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        }
    }
}
```

Trace for sddocname contains 'post'

```
{
    "trace": {
        "children": [
            {
                "message": "No query profile is used"
            },
            {
                "children": [
                    {
                        "children": [
                            {
                                "message": "sc0.num0 search to dispatch: query=[[documentmetastore]:*post*] timeout=19998ms offset=0 hits=0 groupingSessionCache=true sessionId=container.container.0.1580203469223.4390.default grouping=0 :  restrict=[post]"
                            },
                            {
                                "message": [
                                    {
                                        "start_time": "2020-01-28 09:24:29.224 UTC",
                                        "traces": [
                                            {
                                                "timestamp_ms": 0.171641,
                                                "event": "MTF: Start"
                                            },
                                            {
                                                "timestamp_ms": 0.177896,
                                                "event": "MTF: Build query"
                                            },
                                            {
                                                "timestamp_ms": 0.193618,
                                                "event": "MTF: reserve handles"
                                            },
                                            {
                                                "timestamp_ms": 0.558497,
                                                "event": "MTF: Fetch Postings"
                                            },
                                            {
                                                "timestamp_ms": 0.562669,
                                                "event": "MTF: prepareSharedState"
                                            },
                                            {
                                                "timestamp_ms": 0.573859,
                                                "event": "MTF: Complete"
                                            },
                                            {
                                                "timestamp_ms": 0.575737,
                                                "tag": "blueprint",
                                                "optimized": {
                                                    "[type]": "search::queryeval::AndBlueprint",
                                                    "isTermLike": false,
                                                    "estimate": {
                                                        "[type]": "HitEstimate",
                                                        "empty": false,
                                                        "estHits": 32844581,
                                                        "cost_tier": 1,
                                                        "tree_size": 3,
                                                        "allow_termwise_eval": 1
                                                    },
                                                    "sourceId": 4294967295,
                                                    "docid_limit": 32844581,
                                                    "children": {
                                                        "[type]": "std::vector",
                                                        "[0]": {
                                                            "[type]": "search::(anonymous namespace)::AttributeFieldBlueprint",
                                                            "isTermLike": true,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 32844581,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 32844581,
                                                            "fields": {
                                                                "[type]": "FieldList",
                                                                "[0]": {
                                                                    "[type]": "Field",
                                                                    "fieldId": 18,
                                                                    "handle": 0,
                                                                    "isFilter": true
                                                                }
                                                            },
                                                            "attribute": "[documentmetastore]"
                                                        },
                                                        "[1]": {
                                                            "[type]": "proton::documentmetastore::(anonymous namespace)::WhiteListBlueprint",
                                                            "isTermLike": false,
                                                            "estimate": {
                                                                "[type]": "HitEstimate",
                                                                "empty": false,
                                                                "estHits": 32844581,
                                                                "cost_tier": 1,
                                                                "tree_size": 1,
                                                                "allow_termwise_eval": 1
                                                            },
                                                            "sourceId": 4294967295,
                                                            "docid_limit": 32844581
                                                        }
                                                    }
                                                }
                                            },
                                            {
                                                "timestamp_ms": 512.56364,
                                                "tag": "match_threads",
                                                "threads": [
                                                    {
                                                        "traces": [
                                                            {
                                                                "timestamp_ms": 0.630842,
                                                                "event": "Start MatchThread::run"
                                                            },
                                                            {
                                                                "timestamp_ms": 0.713408,
                                                                "event": "Start match and first phase rank"
                                                            },
                                                            {
                                                                "timestamp_ms": 510.085417,
                                                                "event": "Create result set"
                                                            },
                                                            {
                                                                "timestamp_ms": 510.106238,
                                                                "event": "Wait for result processing token"
                                                            },
                                                            {
                                                                "timestamp_ms": 512.154366,
                                                                "event": "Start result processing"
                                                            },
                                                            {
                                                                "timestamp_ms": 512.529474,
                                                                "event": "Start thread merge"
                                                            },
                                                            {
                                                                "timestamp_ms": 512.531487,
                                                                "event": "MatchThread::run Done"
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ],
                                        "distribution-key": 0,
                                        "duration_ms": 512.693125
                                    }
                                ]
                            },
                            {
                                "message": "sc0.num0 dispatch response: Result (0 of total 32844580 hits)"
                            }
                        ]
                    }
                ]
            }
        ]
    },
    "root": {
        "id": "toplevel",
        "relevance": 1,
        "fields": {
            "totalCount": 32844580
        },
        "coverage": {
            "coverage": 100,
            "documents": 32844580,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        }
    }
}
```

**Comment by baldersheim** [comment]
Sorry for radio silence, I have been twisting my head to see where the wholes were in our unit tests. And I found it this morning. https://github.com/vespa-engine/vespa/pull/11996

The short story is that prior to 7.165 there was the bug that you saw initially, where it would incorrectly conclude that there were 0 documents with expert=true. That would be the case after restart, and until a document was fed.

Then in 7.165 this bug was "fixed" by a rewrite for performance reasons. However it introduced another bug that would manifest itself when you added more documents and the field had to "grow". This was fixed in 7.166 without knowing that there was an issue. So in 7.166 things should work fine.

The PR above is to fix the root cause properly and to extend unit test coverage. That will be present 7.169 due out tomorrow.

Sorry for the inconvenience, and thanks a lot for reporting the issue.
